### PR TITLE
Rework player::fire_gun

### PR DESCRIPTION
**Pull request by mugling**
This PR contains a refactor of `player::fire_gun` which also includes the following enhancements:
- ~~Consume moves per number of shots fired in burst mode~~
- Don't add recoil to the player when manually firing a turret
- When calculating point-blank penalty for auxiliary gunmods use total volume for weapon

These refactors are in anticipation of overhauling recoil (see also #15119) and firearm reliability now that we have magazines (see `GAME_BALANCE.md` in #15099).

`recoil_add` will eventually become a member of `player` ~~`item` following the logic that item's (including turrets) recoil and characters are affected by that recoil. This will also help with the `player`/`Character` split as currently `recoil` is only considered for `player`~~.

Does anyone have a good algorithm for how to handle decaying recoil with time?

### CODE REVIEW

**Comment by Coolthulhu** [comment]
> Consume moves per number of shots fired in burst mode

At what multiplier?
I can't find any. Having each shot take the full time would be absolutely crazy for rifles (`15*150 moves=2250 moves = 135 seconds = 2 1/4 minutes` - over 2 minutes of burst fire for 15 bullets?).

> Don't add recoil to the player when manually firing a turret

> recoil_add will eventually become a member of item following the logic that item's (including turrets) recoil and characters are affected by that recoil.

Turrets should have own recoil variable used only for automatic aiming (not player aiming them manually - that should go to player's recoil). Recoil for items themselves doesn't really make sense - it is not the state of the item itself, but the combined state of the shooter, gun and even the target.
It could lead to nonsensical situations like player putting away a shotgun after using it, pulling out an exact copy and having more accuracy with it.

**Comment by mugling** [comment]
> At what multiplier?

Fair point. What would be reasonable?

> It could lead to nonsensical situations like player putting away a shotgun after using it, pulling out an exact copy and having more accuracy with it.

You can already do this (unwielding sets `MIN_RECOIL`). This too needs fixing.

> Turrets should have own recoil variable used only for automatic aiming (not player aiming them manually - that should go to player's recoil).

Why does the player suffer recoil firing a turret mounted half a dozen tiles away since presumably they do so remotely?

> it is not the state of the item itself, but the combined state of the shooter, gun

Indeed, I want to implement a `recoil` property for the item and then adjust `driving_recoil` to become the players overall stability.

**Comment by Coolthulhu** [comment]
> What would be reasonable?

150% of a single shot or something like that. It shouldn't scale too much, especially not with burst size (minigun burst shouldn't last an hour, even if it's 600 bullets).

> You can already do this (unwielding sets MIN_RECOIL). This too needs fixing.

Just capping could be enough.

> Why does the player suffer recoil firing a turret mounted half a dozen tiles away since presumably they do so remotely?

Doesn't really matter since it also resets the recoil before every shot.
Could also reset it after, to represent lack of attention to surroundings.

> Indeed, I want to implement a recoil property for the item and then adjust player::driving_recoil to become the players overall stability.

Why give it to items though?
Turrets will need their separate recoil because turrets aren't items. Weapons won't need their separate recoil because player pulling out or putting one away will always reset it.
Why move the variable to a class that won't use it?

**Comment by mugling** [comment]
> Doesn't really matter since it also resets the recoil before every shot

For manually aimed turrets the player currently accumulates the recoil?

> Just capping could be enough.

Sorry, I don't follow.

> Why give it to items though?
> Turrets will need their separate recoil because turrets aren't items.

Turrets spawn items when they fire and those items also consume ammo and will need to handle reloading. For the purposes of ranged fire turrets effectively are items so the common class would be `item` and not `player`.

**Comment by Malkeus** [comment]
Burst fire is fast. Really really fast. That minigun you mentioned?
2000-6000 rounds per minute. ( https://en.m.wikipedia.org/wiki/Minigun )
These things can spit 600 rounds in 10 seconds.

Sorry for the long link below, I can't get the link from YouTube's page BC
my phone hates me.
Ak47 3 round burst in less than a second (
https://www.google.com/url?sa=t&source=web&rct=j&url=intent://www.youtube.com/watch%3Fv%3DmkrL_sVqLm0%23Intent%3Bscheme%3Dhttp%3Bpackage%3Dcom.google.android.youtube%3BS.browser_fallback_url%3Dhttp%253A%252F%252Fm.youtube.com%252Fwatch%253Fv%253DmkrL_sVqLm0%3BS.android.intent.extra.REFERRER_NAME%3Dhttps%253A%252F%252Fwww.google.com%3Bend&ved=0ahUKEwj98MnlotXKAhXFHh4KHeM2B8IQjjgIJTAC&usg=AFQjCNGrFe22B9sODZjBN8Fg9gMpIuF6DQ&sig2=4-16LEVdWYvNrcD6e7ngEA
)

**Comment by mugling** [comment]
> It shouldn't scale too much, especially not with burst size
> These things can spit 600 rounds in 10 seconds.

On that basis I've reverted 9ae36b7

**Comment by kevingranade** [comment]
Just for a little context about the burst fire, the original idea was that the gun fires X rounds per standard-length action.  Much like single shots, the time consumed is simply the amount of time necessary for the gun to send the given number of rounds downrange, acquiring targets and aiming are handled separately.
Ideally, I'd like the whole burst fire thing to be a bit more flexible, and allow selecting "short burst", "long burst", "3-round burst"/2-round burst (when enabled by the weapon).  To support this the burst stat would probably be swapped out for a "rate-of-fire" stat instead, and the chosen firing mode and possibly a skill check would determine how many rounds are fired (and therefor how long the burst takes).

**Comment by Coolthulhu** [comment]
> For manually aimed turrets the player currently accumulates the recoil?

In your example the turret was remotely controlled. Those don't cause the player to accumulate recoil because they reset it every time.

> Sorry, I don't follow.

Recoil shouldn't be kept for too long. Capping it at some reasonable maximum would prevent unrealistic tactics from FPS like equipping a different gun to reset recoil.
Alternatively some different recoil regeneration values.

> For the purposes of ranged fire turrets effectively are items so the common class would be item and not player.

They're vehicle parts, not items. They spawn items each time they fire, then destroy them after that.
This means that they can't accumulate recoil in the item, but still need own recoil variable.

**Comment by DeadLeaves** [comment]
Isn't this a prime time to allow auxiliary weapons to fire busrt if they have the "BURST" and "BURST_ONLY" flags? That should be a simple addition from what I could tell from the code. It would also fix #14951 trigger happy bug. And most importantly perhaps, it would allow for some really cool aux weapons.

**Comment by mugling** [comment]
@DeadLeaves yes we should correct #14951 in this PR. `BURST_ONLY` remains an outstanding issue, see #12593

**Comment by mugling** [comment]
> They (turrets) spawn items each time they fire, then destroy them after that.

Yes. I was going to adjust that so that a permanent copy is kept around which means we can store a number of other properties in the item (in a particular damage and loaded ammo). Very much interested in any comments or alternatives as fixing the turret reloading is a major chunk of work I need to get sorted.

> In your example the turret was remotely controlled. Those don't cause the player to accumulate recoil because they reset it every time.

`manual_fire_turret` uses the player whilst `automatic_fire_turret` uses a temporary character so at present we have the worst of both worlds (remotely fired turrets add recoil to the player and automatic accumulate none). The fix is in eed044d to not consider turret recoil at all until we have a consistent approach (separate PR).

> Recoil shouldn't be kept for too long. Capping it at some reasonable maximum

So our algorithm needs to entirely decay the recoil within say 200 turns except when the player is still shooting?

> Alternatively some different recoil regeneration values.

I very much like this approach and the model I was working on is that items recoil and the players ability to control that recoil depends upon a number of factors. For example in my mockup low strength or stamina affected how fast recoil was lost from an item.

**Comment by mugling** [comment]
@DeadLeaves Can you check abcebcd fixes #14951 for you.

Presuming we never trigger the `debugmsg` added in 3ad51ea we can then (separate PR) move selection of (maximum) number of rounds to fire out of this function.

**Comment by Coolthulhu** [comment]
> manual_fire_turret uses the player whilst automatic_fire_turret uses a temporary character so at present we have the worst of both worlds (remotely fired turrets add recoil to the player and automatic accumulate none).

Remotely fired turrets do not accumulate recoil because they reset it before every shot.
Player gets the recoil for the last shot/burst, which isn't exactly realistic (remotely using a turret should set the recoil to some default value, to represent losing focus), but it isn't the same as accumulation.
Turrets do accumulate recoil during a single burst, but lose it after that.

I re-read it and I found a bug: manually firing a turret but canceling the attack will drop the recoil to that of the vehicle (which is usually much lower than one for the player).

> I was going to adjust that so that a permanent copy is kept around which means we can store a number of other properties in the item

This isn't wrong per se, it's just that recoil shouldn't really be one of those properties.
Consider all the possible bugs from wielding stuff (like CBM pseudo-wield mechanics), "shared" weapons (turrets), forced unwield and possible others. There's more possible stuff that can go wrong than with player having a static recoil.
Then there's the fact that recoil of an unwielded item is itself a nonsensical value.
Also consider the other fake shooter: monster attacks.

Instead storing it as a turret value and supplying to fire function would be more proper.
If the number of extra parameters grew too big, you could have a `shooter` structure, similar to that of a `projectile` to avoid passing each parameter explicitly. This structure could be stored or rebuilt, depending on how much it changed.

Now that I think about it, your idea with turning driving recoil into general stability makes sense, but the "other recoil" should represent not the stability of the gun, but how much it points at the current target.

> So our algorithm needs to entirely decay the recoil within say 200 turns except when the player is still shooting?

Something like that.
In general I see recoil like accuracy drop in FPS: increases when you move, increases a lot when you shoot, decreases when you stand still, decreases heavily when you aim. No action that doesn't involve aiming should decrease it faster than standing still.
For turrets, player using a turret should reset turret's aim, meaning that it should not cause it to accumulate recoil for the next turn, but it should lose its own aim (once turret aiming becomes a thing, that is). Turrets are too stable to suffer recoil for more than 6 seconds.
> For example in my mockup low strength or stamina affected how fast recoil was lost from an item.

Stamina makes sense because shaking hands etc., but I'd keep strength to just lowering recoil during shooting, not after it. This effect could be buffed because currently it is very minor - changing it to a scaling one (division instead of subtraction) would be helpful at making (high recoil) bursts become useful for the player.
When in doubt, nerf energy weapons and buff stuff that consumes a lot of expensive ammo.

**Comment by mugling** [comment]
Ok I've changed (b9dfb0e) the order of arguments in `recoil_add` and code comments with a view to later moving this to be part of `player`

> Instead storing it as a turret value and supplying to fire function would be more proper.
> If the number of extra parameters grew too big, you could have a shooter structure, similar to that of a projectile to avoid passing each parameter explicitly. This structure could be stored or rebuilt, depending on how much it changed.

Sounds like the most appropriate plan (but separate PR)

**Comment by Coolthulhu** [comment]
Did you remove that hardcoded rifle range bonus thing? Why?
It was quite ugly so I won't defend having it in current form, but it was a sensible boost to sniping - sniping is currently rather bad.

**Comment by mugling** [comment]
> It was quite ugly so I won't defend having it in current form

That exact reason. I want to replace it with something else but I'm not sure what yet

Additionally I want to make the point-blank penalty much harsher so that pistols are much better weapons inside buildings.

**Comment by Coolthulhu** [comment]
> Additionally I want to make the point-blank penalty much harsher so that pistols are much better weapons inside buildings.

Don't
I mean, not just linearly.
Pistols are already better than rifles when you consider the time per shot - pistols shoot almost 2x as fast at 0 skill and 3x as fast at skill 7+.
If you want to nerf point blank shooting, do it by allowing zombies to "dodge" the shot (not Matrix style, but during the aiming) rather than by flat penalty for the shooter.
I tried implementing it in `Creature::deal_projectile_attack` but I didn't complete it (it only allows dodging slow projectiles).

File: src/ranged.cpp [hunk 1]
```diff
@@ -342,36 +342,32 @@ void player::fire_gun( const tripoint &target, bool burst )
...
     }
     const skill_id skill_used = gun.gun_skill();
 
-    if (has_trait("TRIGGERHAPPY") && one_in(30)) {
+    if( has_trait("TRIGGERHAPPY") && one_in( 30 ) ) {
         burst = true;
     }
     if( burst && gun.burst_size() < 2 ) {
-        burst = false; // Can't burst fire a semi-auto
+        debugmsg( "Tried to burst fire a semi-auto" );
```
Comment (Coolthulhu) [inline src/ranged.cpp]: This causes debugmsgs when using a semi-auto weapon with TRIGGERHAPPY.

Comment (mugling) [inline src/ranged.cpp]: Good find. Fixed in e0fd4e4

**Pull request merged**